Sources/MapboxMaps/Foundation/ObservableValue.swift

```diff
@@ -1,9 +1,99 @@
+import Foundation
+
+/// Wrapper for ``ObservableValue`` to throttle its value updates.
+/// Republishes only the latest event received during the time window interval.
+internal final class Throttle<Value> where Value: Equatable {
+    internal var value: Value?
+
+    private var subscriptions = [BlockSubscription<Value>]()
+    private var latestValue: Value? { observableValue.value }
+
+    private let observableValue: ObservableValue<Value>
+    private let dispatchQueue: DispatchQueueProtocol
+    private let windowDuration: TimeInterval
+    private var cancelToken: Cancelable?
+    private var item: DispatchWorkItem?
+
+    deinit {
+        cancelToken?.cancel()
+    }
+
+    internal init(
+        value: ObservableValue<Value> = .init(),
+        windowDuration: TimeInterval,
+        dispatchQueue: DispatchQueueProtocol = DispatchQueue.main
+    ) {
+        self.observableValue = value
+        self.dispatchQueue = dispatchQueue
+        self.windowDuration = windowDuration
+
+        cancelToken = observableValue.observe { [weak self] newValue in
+            if self?.value == nil { // first update
+                self?.notifyImmediately(with: newValue)
+            } else {
+                self?.onValueUpdated(newValue: newValue)
```

Comment:
what would happen if the 2nd value is emitted after a duration that is bigger than `windowDuration` since 1st value, do we emit 2nd value immediately or we wait for extra `windowDuration`to pass?

![diagram](https://user-images.githubusercontent.com/30261043/203946752-786621e3-a0ea-48cd-9b36-88d3fb5f04f6.png)